autode/reactions/multistep.py

```diff
@@ -9,60 +9,207 @@
+        number of atoms is conserved throughout the reaction. Allows for
+        multistep reactions where molecules are added e.g. catalytic cycles
+        """
+
+        for i in range(self.n_steps - 1):
+            j = i + 1   # Index of the next step in the sequence of reactions
+
+            if self.reactions[j].has_identical_composition_as(self.reactions[i]):
+                continue
+
+            if self._adds_molecule(i, j):
+                self._add_mol_to_reacs_prods(mol=self._added_molecule(i, j),
+                                             to_idx=j)
 
-            if i == 0:
-                # First reaction has no previous products
-                calculate(reaction=r, prev_products=[])
+            elif self._adds_molecule(j, i):
+                self._add_mol_to_reacs_prods(mol=self._added_molecule(j, i),
+                                             from_idx=j)
 
             else:
-                calculate(reaction=r, prev_products=self.reactions[i-1].prods)
+                raise RuntimeError('Failed to balance the multistep reaction. '
+                                   'No reactants present as products of a '
+                                   'previous step.')
+
+        return None
+
+    def _adds_molecule(self,
+                       step_idx:      int,
+                       next_step_idx: int
+                       ) -> bool:
+        """Does reaction 2 have more molecules(atoms) than reaction 1"""
+
+        def total_n_atoms(rxn):
+            return sum(m.n_atoms for m in rxn.reacs)
+
+        return (total_n_atoms(self.reactions[step_idx])
+                < total_n_atoms(self.reactions[next_step_idx]))
+
+    def _add_mol_to_reacs_prods(self, mol, from_idx=0, to_idx=None):
+        """Add a molecule to both reactants and products"""
+        if to_idx is None:
+            to_idx = self.n_steps
+
+        for reaction in self.reactions[from_idx:to_idx]:
+            reaction.reacs.append(mol.to_reactant())
+            reaction.prods.append(mol.to_product())
+
+        return None
+
+    def _added_molecule(self,
+                        step_idx:      int,
+                        next_step_idx: int
+                        ) -> 'autode.species.molecule.Molecule':
+        r"""
+        Extract the added molecule going from a step to the next. For example::
+
+            A + B -> C
+            C + D -> E
+
+        the added molecule would be D. The balanced sequence is then:
+        A + B + D -> C + D -> E
+        -----------------------------------------------------------------------
+        Arguments:
+            step_idx:
+            next_step_idx:
+
+        Returns:
+            (autode.species.molecule.Molecule):
+
+        Raises:
+            (RuntimeError):
+        """
+        prods = self.reactions[step_idx].prods
+
+        for mol in self.reactions[next_step_idx].reacs:
+            if any(mol.atomic_symbols == p.atomic_symbols for p in prods):
```

Comment:
I don't think this works in the case of
```
A + B -> C
C + C -> D
```
?

Reply:
That is an exceedingly good point. Would need to consider `2(A+B) -> 2C -> D` scenarios and generalisations... hmm. I don't know of any reactions like that so for the time being I'm okay for it to just raise a `RuntimeError` but that's worth adding in the docstring I guess(?)

Reply:
yep that sounds sensible to me :)